### PR TITLE
[linter] Updating RuleID and generalizing related description missing rules

### LINKED ISSUE #1919: ParameterDescriptionRequired: Message needs to be improved, ID updated. Description related rules should be combined/more consistent.
ID: M4000
Improve message to include property/parameter/entity name.

Evaluate whether other description rules should be combined, or more consistent. Like ModelTypeIncomplete, OperationDescriptionRequired, DescriptiveDescriptionRequired.

The above share the ID: M4000

### PR DESCRIPTION
**Pull request by vishrutshah**
Fixes #1919 
Fixes https://github.com/Azure/autorest/issues/2044

### CODE REVIEW

File: src/modeler/AutoRest.Swagger/Validation/ModelTypeIncomplete.cs [hunk 1]
```diff
@@ -8,30 +8,23 @@
...
+        /// <returns><c>true</c> if model contains description, <c>false</c> otherwise</returns>
         public override IEnumerable<ValidationMessage> GetValidationMessages(Schema schema, RuleContext context)
         {
-            if (schema != null && string.IsNullOrEmpty(schema.Reference) && schema.RepresentsCompositeType())
+            if (string.IsNullOrWhiteSpace(schema?.Reference) && schema.RepresentsCompositeType())
             {
                 if (schema.Description == null)
                 {
-                    yield return new ValidationMessage(new FileObjectPath(context.File, context.Path), this, "description");
+                    yield return new ValidationMessage(new FileObjectPath(context.File, context.Path), this, "model type");
```
Comment (veronicagg) [inline src/modeler/AutoRest.Swagger/Validation/ModelTypeIncomplete.cs]: could we get the name of the model type in the message?

File: src/modeler/AutoRest.Swagger/Validation/ModelTypeIncomplete.cs [hunk 2]
```diff
@@ -8,30 +8,23 @@
...
+        /// </summary>
+        /// <param name="schema">Schema being validated</param>
+        /// <param name="context">Rule context</param>
+        /// <returns><c>true</c> if model contains description, <c>false</c> otherwise</returns>
         public override IEnumerable<ValidationMessage> GetValidationMessages(Schema schema, RuleContext context)
         {
-            if (schema != null && string.IsNullOrEmpty(schema.Reference) && schema.RepresentsCompositeType())
+            if (string.IsNullOrWhiteSpace(schema?.Reference) && schema.RepresentsCompositeType())
             {
                 if (schema.Description == null)
```
Comment (veronicagg) [inline src/modeler/AutoRest.Swagger/Validation/ModelTypeIncomplete.cs]: should this condition also consider if description is empty or whitespace, similar to the other ones?

File: src/modeler/AutoRest.Swagger/Validation/OperationDescriptionRequired.cs [hunk 3]
```diff
@@ -1,34 +1,24 @@
...
-        /// </summary>
-        public override Category Severity => Category.Warning;
+        /// <param name="entity">Entity being validated</param>
+        /// <param name="context">Rule context</param>
+        /// <param name="formatParameters">formatted parameters</param>
+        /// <returns><c>true</c> if entity contains description, <c>false</c> otherwise</returns>
+        public override bool IsValid(Operation entity, RuleContext context, out object[] formatParameters)
+        {
+            formatParameters = new string[] { string.Format("operation '{0}'", entity.OperationId) };
+            return entity?.Description != null;
```
Comment (veronicagg) [inline src/modeler/AutoRest.Swagger/Validation/OperationDescriptionRequired.cs]: should this condition be the same as the one for "parameters" "return entity?.Description != null || !string.IsNullOrWhiteSpace(entity.Reference);"

File: src/modeler/AutoRest.Swagger/Validation/ParameterDescriptionRequired.cs [hunk 4]
```diff
@@ -1,34 +1,24 @@
...
-        /// The severity of this message (ie, debug/info/warning/error/fatal, etc)
-        /// </summary>
-        public override Category Severity => Category.Warning;
+        /// <param name="entity">Entity being validated</param>
+        /// <param name="context">Rule context</param>
+        /// <param name="formatParameters">formatted parameters</param>
+        /// <returns><c>true</c> if entity contains description, <c>false</c> otherwise</returns>
+        public override bool IsValid(SwaggerParameter entity, RuleContext context, out object[] formatParameters)
+        {
+            formatParameters = new string[] { "parameter" };
```
Comment (veronicagg) [inline src/modeler/AutoRest.Swagger/Validation/ParameterDescriptionRequired.cs]: could we get the parameter name in the message?

File: src/modeler/AutoRest.Swagger/Validation/OperationDescriptionRequired.cs (hunk 3 above)
Comment (vishrutshah) [inline src/modeler/AutoRest.Swagger/Validation/OperationDescriptionRequired.cs]: I am not sure what is `entity.Reference` in the context of the operation. I should definitely check for whitespace though

Comment (vishrutshah) [inline src/modeler/AutoRest.Swagger/Validation/OperationDescriptionRequired.cs]: Updated it to ` return !string.IsNullOrWhiteSpace(entity?.Description);`

File: src/modeler/AutoRest.Swagger/Validation/ModelTypeIncomplete.cs [hunk 5]
```diff
@@ -8,30 +8,23 @@
...
+        /// <summary>
+        /// Validates model for description property
+        /// </summary>
+        /// <param name="schema">Schema being validated</param>
+        /// <param name="context">Rule context</param>
+        /// <returns><c>true</c> if model contains description, <c>false</c> otherwise</returns>
         public override IEnumerable<ValidationMessage> GetValidationMessages(Schema schema, RuleContext context)
         {
-            if (schema != null && string.IsNullOrEmpty(schema.Reference) && schema.RepresentsCompositeType())
+            if (string.IsNullOrWhiteSpace(schema?.Reference) && schema.RepresentsCompositeType())
```
Comment (dsgouda) [inline src/modeler/AutoRest.Swagger/Validation/ModelTypeIncomplete.cs]: You will still want to do ```schema?.RepresentsCompositeType() ==true```

File: src/modeler/AutoRest.Swagger/Validation/ModelTypeIncomplete.cs [hunk 6]
```diff
@@ -8,30 +8,23 @@
...
             {
-                if (schema.Description == null)
+                if (string.IsNullOrWhiteSpace(schema.Description))
```
Comment (dsgouda) [inline src/modeler/AutoRest.Swagger/Validation/ModelTypeIncomplete.cs]: ```schema?.Description```

File: src/modeler/AutoRest.Swagger/Validation/ParameterDescriptionRequired.cs (hunk 4 above)
Comment (dsgouda) [inline src/modeler/AutoRest.Swagger/Validation/ParameterDescriptionRequired.cs]: I'm guessing you're using this validation rule multiple times in the spec. It's usually more efficient to set this as a readonly static variable instead so we don't create the same object multiple times unnecessarily

File: src/modeler/AutoRest.Swagger/Validation/ModelTypeIncomplete.cs (hunk 2 above)
Comment (vishrutshah) [inline src/modeler/AutoRest.Swagger/Validation/ModelTypeIncomplete.cs]: fixed

File: src/modeler/AutoRest.Swagger/Validation/ModelTypeIncomplete.cs (hunk 6 above)
Comment (vishrutshah) [inline src/modeler/AutoRest.Swagger/Validation/ModelTypeIncomplete.cs]: After having 

`if (string.IsNullOrWhiteSpace(schema?.Reference) && schema?.RepresentsCompositeType() == true)` will i get null schema inside?

File: src/modeler/AutoRest.Swagger/Validation/ParameterDescriptionRequired.cs (hunk 4 above)
Comment (vishrutshah) [inline src/modeler/AutoRest.Swagger/Validation/ParameterDescriptionRequired.cs]: sure

**Comment by vishrutshah** [comment]
@veronicagg I looked at the PR and saw that we were not applying the rule for the properties of the model. Now it's corrected by applying `ModelTypeIncoplete` rule to model properties as well. Also added test in the same swagger. 

Here is the output for that PR with removed one description inside mode.

```
  {
    "type": "Warning",
    "code": "ModelTypeIncomplete",
    "message": "This 'availableClusterVersions' model/property lacks 'description' property. Consider adding a 'description' element. Accurate description is essential for maintaining reference documentation.",
    "jsonref": "file:///D:/Repos/azure-rest-api-specs/arm-servicefabric/2016-09-01/swagger/servicefabric.json#/definitions/ClusterProperties/properties",
    "json-path": "file:///D:/Repos/azure-rest-api-specs/arm-servicefabric/2016-09-01/swagger/servicefabric.json#/definitions/ClusterProperties/properties",
    "id": "M4000",
    "validationCategory": "SDKViolation",
    "providerNamespace": null,
    "resourceType": null
  },
  {
    "type": "Warning",
    "code": "ModelTypeIncomplete",
    "message": "This 'value' model/property lacks 'description' property. Consider adding a 'description' element. Accurate description is essential for maintaining reference documentation.",
    "jsonref": "file:///D:/Repos/azure-rest-api-specs/arm-servicefabric/2016-09-01/swagger/servicefabric.json#/definitions/ClusterListResult/properties",
    "json-path": "file:///D:/Repos/azure-rest-api-specs/arm-servicefabric/2016-09-01/swagger/servicefabric.json#/definitions/ClusterListResult/properties",
    "id": "M4000",
    "validationCategory": "SDKViolation",
    "providerNamespace": null,
    "resourceType": null
  }
```

Also addressed the comments. @veronicagg / @dsgouda please look at the PR again. Thanks

File: src/modeler/AutoRest.Swagger/Validation/ModelTypeIncomplete.cs (hunk 6 above)
Comment (dsgouda) [inline src/modeler/AutoRest.Swagger/Validation/ModelTypeIncomplete.cs]: Ahh, fair point

File: src/modeler/AutoRest.Swagger/Validation/DescriptionRequired.cs [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+        /// </summary>
+        public override ValidationCategory ValidationCategory => ValidationCategory.SDKViolation;
+
+        /// <summary>
+        /// The template message for this Rule. 
+        /// </summary>
+        /// <remarks>
+        /// This may contain placeholders '{0}' for parameterized messages.
+        /// </remarks>
+        public override string MessageTemplate => Resources.MissingDescription;
```
Comment (dsgouda) [inline src/modeler/AutoRest.Swagger/Validation/DescriptionRequired.cs]: If this is going to get overridden in the base classes anyway, no need to declare here

File: src/modeler/AutoRest.Swagger/Validation/ModelTypeIncomplete.cs [hunk 8]
```diff
@@ -8,30 +8,30 @@
...
+        /// <returns><c>true</c> if model contains description, <c>false</c> otherwise</returns>
+        public override IEnumerable<ValidationMessage> GetValidationMessages(Dictionary<string, Schema> definitions, RuleContext context)
         {
-            if (schema != null && string.IsNullOrEmpty(schema.Reference) && schema.RepresentsCompositeType())
+            foreach (KeyValuePair<string, Schema> definition in definitions)
             {
-                if (schema.Description == null)
+                string key = definition.Key;
+                Schema schema = definition.Value;
+                if (string.IsNullOrWhiteSpace(schema?.Reference))
```
Comment (dsgouda) [inline src/modeler/AutoRest.Swagger/Validation/ModelTypeIncomplete.cs]: Why do we need to check for ```schema.Reference``` here?

File: src/modeler/AutoRest.Swagger/Validation/ParameterDescriptionRequired.cs [hunk 9]
```diff
@@ -1,34 +1,26 @@
...
         /// </summary>
-        public override Category Severity => Category.Warning;
+        /// <param name="entity">Entity being validated</param>
+        /// <param name="context">Rule context</param>
+        /// <param name="formatParameters">formatted parameters</param>
+        /// <returns><c>true</c> if entity contains description, <c>false</c> otherwise</returns>
+        public override bool IsValid(SwaggerParameter entity, RuleContext context, out object[] formatParameters)
+        {
+            formatParameters = new string[] { string.Format(ParameterTypeFormatter, entity.Name) };
+            return !string.IsNullOrWhiteSpace(entity?.Description) || !string.IsNullOrWhiteSpace(entity.Reference);
```
Comment (dsgouda) [inline src/modeler/AutoRest.Swagger/Validation/ParameterDescriptionRequired.cs]: nit:The rule won't be called for a null entity, you can say ```entity.Description```

File: src/core/AutoRest.Core/Properties/Resources.resx [hunk 10]
```diff
@@ -283,7 +283,7 @@ Modelers:
     <value>'{0}' is not a known format.</value>
   </data>
   <data name="MissingDescription" xml:space="preserve">
-    <value>Consider adding a 'description' element, essential for maintaining reference documentation.</value>
+    <value>This {0} lacks 'description' property. Consider adding a 'description' element. Accurate description is essential for maintaining reference documentation.</value>
```
Comment (vishrutshah) [inline src/core/AutoRest.Core/Properties/Resources.resx]: remove `this`

File: src/modeler/AutoRest.Swagger/Validation/OperationDescriptionRequired.cs [hunk 11]
```diff
@@ -1,34 +1,24 @@
...
-        /// The severity of this message (ie, debug/info/warning/error/fatal, etc)
-        /// </summary>
-        public override Category Severity => Category.Warning;
+        /// <param name="entity">Entity being validated</param>
+        /// <param name="context">Rule context</param>
+        /// <param name="formatParameters">formatted parameters</param>
+        /// <returns><c>true</c> if entity contains description, <c>false</c> otherwise</returns>
+        public override bool IsValid(Operation entity, RuleContext context, out object[] formatParameters)
+        {
+            formatParameters = new string[] { string.Format("operation '{0}'", entity.OperationId) };
```
Comment (vishrutshah) [inline src/modeler/AutoRest.Swagger/Validation/OperationDescriptionRequired.cs]: caps on `O`

File: src/modeler/AutoRest.Swagger/Validation/ParameterDescriptionRequired.cs (hunk 9 above)
Comment (vishrutshah) [inline src/modeler/AutoRest.Swagger/Validation/ParameterDescriptionRequired.cs]: `Reference` cannot be in `SwaggerParameter` schema from `http://swagger.io/specification/#parameterObject`

File: src/modeler/AutoRest.Swagger/Validation/DescriptionRequired.cs (hunk 7 above)
Comment (vishrutshah) [inline src/modeler/AutoRest.Swagger/Validation/DescriptionRequired.cs]: as discussed offline, this is the base class

**Comment by vishrutshah** [comment]
@azuresdkci Test this please

File: src/modeler/AutoRest.Swagger/Validation/ParameterDescriptionRequired.cs (hunk 9 above)
Comment (vishrutshah) [inline src/modeler/AutoRest.Swagger/Validation/ParameterDescriptionRequired.cs]: @veronicagg I found that `parameters` in the operations can have refs. So i cannot remove this. 

ex: https://github.com/Azure/autorest/blob/master/src/modeler/AutoRest.Swagger.Tests/Resource/Swagger/Validation/positive/clean-complex-spec.json#L166

**Comment by vishrutshah** [comment]
@azuresdkci Test this please

File: src/modeler/AutoRest.Swagger/Validation/OperationDescriptionRequired.cs [hunk 12]
```diff
@@ -1,34 +1,24 @@
...
-        /// </summary>
-        public override Category Severity => Category.Warning;
+        /// <param name="entity">Entity being validated</param>
+        /// <param name="context">Rule context</param>
+        /// <param name="formatParameters">formatted parameters</param>
+        /// <returns><c>true</c> if entity contains description, <c>false</c> otherwise</returns>
+        public override bool IsValid(Operation entity, RuleContext context, out object[] formatParameters)
+        {
+            formatParameters = new string[] { string.Format("Operation '{0}'", entity.OperationId) };
+            return !string.IsNullOrWhiteSpace(entity?.Description);
```
Comment (dsgouda) [inline src/modeler/AutoRest.Swagger/Validation/OperationDescriptionRequired.cs]: Guess you missed ```entity?.Description```

**Comment by vishrutshah** [comment]
Thanks for the review!!

**Pull request merged**